netbench-orchestrator/src/russula/mod.rs

```diff
@@ -0,0 +1,170 @@
+use std::{collections::BTreeSet, net::SocketAddr};
+use tokio::net::TcpStream;
+use tracing::{error, info};
+
+mod error;
+mod event;
+mod network_utils;
+mod protocol;
+mod states;
+
+use error::{RussulaError, RussulaResult};
+use protocol::Protocol;
+
+const CONNENT_RETRY_ATTEMPT: usize = 10;
+
+#[derive(Debug, Copy, Clone)]
+pub enum ProtocolState {
+    /// The endpoint has established connection with its peer and
+    /// is ready to make progress.
+    Ready,
+
+    /// Indicates the protocol's terminal state.
+    Done,
+
+    /// Indicates that worker are running and accepting work.
+    ///
+    /// For netbench this state be used to confirm that all servers are
+    /// running and accepting connection before starting netbench clients.
+    /// Should only be called by Coordinators.
+    WorkerRunning,
+}
+
+/// An instance of a protocol with an established connection to its peer.
+struct ProtocolInstance<P: Protocol> {
+    pub addr: SocketAddr,
+    pub stream: TcpStream,
+    pub protocol: P,
+}
+
+/// A Russula Endpoint.
+///
+/// An Endpoint can be of type Coordinator or Worker. A Coordinator can
+/// be used to synchronize multiple workers across different hosts. A Worker
+/// communicates with a Coordinator to make progress.
+pub struct Endpoint<P: Protocol> {
+    /// List of protocol instance to synchronize with.
+    instance_list: Vec<ProtocolInstance<P>>,
+
+    /// Polling frequency when trying to make progress.
+    poll_delay: Duration,
+}
+
+impl<P: Protocol + Send> Endpoint<P> {
+    pub async fn run_till(&mut self, state: ProtocolState) -> RussulaResult<()> {
+        while self.poll_state(state).await?.is_pending() {
+            tokio::time::sleep(self.poll_delay).await;
+        }
+
+        Ok(())
+    }
+
+    pub async fn poll_state(&mut self, state: ProtocolState) -> RussulaResult<Poll<()>> {
+        // Poll each peer protocol instance.
+        //
+        // If the peer is already in the desired state then this should be a noop.
+        for peer in self.instance_list.iter_mut() {
+            if let Err(err) = peer.protocol.poll_state(&mut peer.stream, state).await {
+                if err.is_fatal() {
+                    error!("{} {}", err, peer.addr);
+                    panic!("{} {}", err, peer.addr);
+                }
+            }
+        }
+
+        // Check that all instances are at the desired state.
+        let poll = if self.is_state(state) {
+            Poll::Ready(())
+        } else {
+            Poll::Pending
+        };
+        Ok(poll)
+    }
+
+    /// Check if all instances are at the desired state
+    fn is_state(&self, state: ProtocolState) -> bool {
+        for peer in self.instance_list.iter() {
+            // All instance must be at the desired state
+            if !peer.protocol.is_state(state) {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+type SockProtocol<P> = (SocketAddr, P);
+pub struct RussulaBuilder<P: Protocol> {
```

Comment:
Russula is agnostic over netbench. In future PRs we will see that you can implement Netbench specific protocols in Russula so I would like to keep the naming separate. There is a description of what Russula is in the Readme. If some part of that is not clear then I can expand on that.

Reply:
The "establish a usable connection during the build step" is not uncommon and we use it in s2n-quic. A [Client::connect](https://docs.rs/s2n-quic/latest/s2n_quic/client/struct.Client.html#method.connect) returns a [Connection](https://docs.rs/s2n-quic/latest/s2n_quic/connection/struct.Connection.html) which can do more interesting stuff.

Having separate  pre-connect/post-connect struct/state is important to prevent someone from trying to `poll_until(state)` before establishing a connection with the peer.

I have added some documentation to this and renamed it to `Builder`.